runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -973,65 +1277,313 @@ MM_SchedulingDelegate::calculateAutomaticGMPIntermission(MM_EnvironmentVLHGC *en
+	}
+
+	intptr_t heapNotFullyExpandedRecommendation = 0;
+	intptr_t heapFullyExpandedRecommendation = 0;
+
+	if (0.0 == heapPercentExpandedAboveThreshold) {
+		heapNotFullyExpandedRecommendation = calculateEdenChangeHeapNotFullyExpanded(env);
+	} else if (0.0 < heapPercentExpandedAboveThreshold) {
+		/** 
+		 * If heap is almost fully expanded, start to consider the recommendation given from moveTowardRecommendedEdenForExpandedHeap().
+		 */
+		if (globalSweepHappened) {
+			heapFullyExpandedRecommendation = moveTowardRecommendedEdenForExpandedHeap(env, 0.5);
+			heapNotFullyExpandedRecommendation = calculateEdenChangeHeapNotFullyExpanded(env);
+		} else if (0 == _pgcCountSinceGMPEnd % consecutivePGCToChangeEden){
+			/** 
+	 		 * Every consecutivePGCToChangeEden number of PGC's, check to see if eden size should change. 
+			 * This allows PGC to do some defragmentation work to finish, and for some statistics to stabilize
+ 			 */
+			heapFullyExpandedRecommendation = moveTowardRecommendedEdenForExpandedHeap(env, 0.25);
+			heapNotFullyExpandedRecommendation = calculateEdenChangeHeapNotFullyExpanded(env);
+		}
+	}
+
+	if (globalSweepHappened) {
+		resetPgcTimeStatistics(env);
+	}
+
+	Trc_MM_SchedulingDelegate_checkEdenSizeAfterPgc(env->getLanguageVMThread(), heapNotFullyExpandedRecommendation, heapFullyExpandedRecommendation, ratioOfHeapExpanded);
+
+	/** 
+	 * Once ratioOfHeapExpanded is over heapFullyExpandedThreshold, start to blend the recommendations provided by 
+	 * heapFullyExpandedRecommendation and heapNotFullyExpandedRecommendation, based off of how closely heap is to being fully expanded.
+	 * We do this to ensure a smoother transition between the 2 recommendations, which could potentially be pulling in opposite directions.
+	 * If the heap is 100% expanded for example, only care about heapFullyExpandedRecommendation, whereas if heap is not close to being fully expanded, 
+	 * only consider heapNotFullyExpandedRecommendation for purpose of eden resizing.
+	 */
+	double heapFullyExpandedSuggestionWeight = heapPercentExpandedAboveThreshold / (1.0 - heapFullyExpandedThreshold);
+	_edenSizeFactor += (intptr_t)MM_Math::weightedAverage((double)heapFullyExpandedRecommendation, (double)heapNotFullyExpandedRecommendation, heapFullyExpandedSuggestionWeight);
+}
+
+intptr_t
+MM_SchedulingDelegate::calculateEdenChangeHeapNotFullyExpanded(MM_EnvironmentVLHGC *env)
+{
+	intptr_t edenRegionChange = 0;
+	intptr_t edenChangeMagnitude = (intptr_t)ceil((0.03 * getIdealEdenSizeInBytes(env)) / _regionManager->getRegionSize());
+
+	double hybridEdenOverhead = calculateHybridEdenOverhead(env, (uintptr_t)_historicalPartialGCTime, _partialGcOverhead, false);
+
+	Trc_MM_SchedulingDelegate_calculateEdenChangeHeapNotFullyExpanded(env->getLanguageVMThread(), hybridEdenOverhead, (uintptr_t)_historicalPartialGCTime, mapPgcPauseOverheadToPgcCPUOverhead(env, (uintptr_t)_historicalPartialGCTime, false));
```

Comment:
I'd move this trace point inside calculateHybridEdenOverhead (and rename it accordingly) to avoid double calculation of mapPgcPauseOverheadToPgcCPUOverhead. Then you can expand it to report heapFullyExpanded parameter, as well (you can 'convert' bool to strings easily with '?:' conditional operator) and make it useful in expanded case, too.

Reply:
If I move this tracepoint inside of `calculateHybridEdenOverhead()`, then it will be printed 100 or so times. This is because `calculateRecommendedEdenChangeForExpandedHeap()` calls `calculateHybridEdenOverhead()` when it is trying to find the best value for eden size. 

That being said, I can add a parameter in `calculateHybridEdenOverhead()` which indicates if the tracepoint should be run.  I agree that recalculating `mapPgcPauseOverheadToPgcCPUOverhead ()` is a bit wasteful, but the alternative (passing in a parameter indicating whether or not the tracepoint should run), might be a little unorthodox (I haven't seen any other functions do this, but It certainly might exist somewhere). Up to you which approach is better.